src/com/amazon/ion/impl/lite/IonDatagramLite.java

```diff
@@ -300,31 +300,24 @@ public boolean addAll(int index, Collection<? extends IonValue> c)
         return changed;
     }
 
+    @Override
+    int hashSignature() {
+        return HASH_SIGNATURE;
+    }
+
     @Override
     public int hashCode() {
         int prime  = 8191;
         int result = HASH_SIGNATURE;
 
-        if (!isNullValue()) {
-            // As we are a datagram then the children need to resolve their own symbol tables -
-            // so we use the 'top level' #hashCode() which will force each child to resolve it's
-            // own symbol table.
-            for (IonValue v : this) {
-                result = prime * result + v.hashCode();
-                // mixing at each step to make the hash code order-dependent
-                result ^= (result << 29) ^ (result >> 3);
-            }
+        for (IonValue v : this) {
+            result = prime * result + v.hashCode();
+            // mixing at each step to make the hash code order-dependent
+            result ^= (result << 29) ^ (result >> 3);
         }
```

Comment:
I agree, I don't particularly like any of the manual hashCode calculations. My goal for this change, however, is to avoid changing any of the hash codes returned by `IonValue.hashCode`, so I copied the logic as-is. If we're going to change the hash codes that are returned, I'd want to do a more thorough consideration of any implications that may have, separate from the immediate need being addressed by this PR. Would you accept an issue for this?

Reply:
https://github.com/amazon-ion/ion-java/issues/559